### PR TITLE
[16.0][MIG] base_business_document_import

### PR DESCRIPTION
**Pull request by ghost**
Depends on:
- [x] #723

### CODE REVIEW

**Comment by StephaneMangin** [comment]
cc @gbrito-erpopen Could you check for your dependencies please ? Once fixed, we could move forward.

**Comment by ghost** [comment]
@StephaneMangin 
There is an open pr for the dependency #723

**Comment by simahawk** [comment]
> @StephaneMangin There is an open pr for the dependency #723

I added it to the description, can you take of this next time? :pray: 
Also, you can fix tests here by including a pending PR by adding a separated commit w/ this change in `tests-requirements.txt`:

```
odoo-addon-pdf_helper @ git+https://github.com/OCA/edi@refs/pull/723/head#subdirectory=setup/pdf_helper
```

**Comment by ghost** [comment]
> > @StephaneMangin There is an open pr for the dependency #723
> 
> I added it to the description, can you take of this next time? pray Also, you can fix tests here by including a pending PR by adding a separated commit w/ this change in `tests-requirements.txt`:
> 
> ```
> odoo-addon-pdf_helper @ git+https://github.com/OCA/edi@refs/pull/723/head#subdirectory=setup/pdf_helper
> ```

Good tip, I will for sure use it

**Comment by ghost** [comment]
@simahawk I added the test-requirements file, the tests pass, except the unreleased dependencies. is there something I can do to make sure it is also green?

**Comment by simahawk** [comment]
> @simahawk I added the test-requirements file, the tests pass, except the unreleased dependencies. is there something I can do to make sure it is also green?

No, that's normal. We cannot merge a PR if it misses another PR to work.
This is also the nice thing of having the pending merge added in tests-requirements.txt:  you get automatic check for it.
TIP: always put such change in a different commit (eg: `Add TMP test dependency`) so that:
1. is clearly visible in the commit list
2. to remove it, you just discard the commit w/o rewriting the last good one

**Comment by ghost** [comment]
Ah, i see i will make the change in this PR, i add it together. Thanks

**Comment by bosd** [comment]
This one could use a bit more test coverage.

File: base_business_document_import/models/business_document_import.py [hunk 1]
```diff
@@ -0,0 +1,1313 @@
...
+                    )
+            else:
+                chatter_msg.append(
+                    _("Missing VAT number on company '%s'") % company.display_name
+                )
+
+    def get_xml_files_from_pdf(self, pdf_file):
+        """Returns a dict with key = filename, value = XML file obj"""
+        logger.warning(
+            "`get_xml_files_from_pdf` deprecated: use `pdf.helper.pdf_get_xml_files`"
```
Comment (simahawk) [inline base_business_document_import/models/business_document_import.py:1373]: this method should be removed

**Comment by ghost** [comment]
> This one could use a bit more test coverage.

Sure, I will add more tests to it

**Review by sbejaoui — APPROVED** [review]

**Review by StephaneMangin — APPROVED** [review]

File: base_business_document_import/models/business_document_import.py [hunk 2]
```diff
@@ -0,0 +1,1320 @@
...
+                    ).format(
+                        record=record.display_name,
+                        record_model=record._name,
+                        target_model=model._name,
+                    )
+                )
+        if data_dict.get("id"):
+            record = False
+            try:
+                record = model.browse(data_dict["id"])
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Simper way to do it: `record = model.browse(data_dict["id"]).exists()`
(so the next line with `record.create_date` become useless.)

File: base_business_document_import/models/business_document_import.py [hunk 3]
```diff
@@ -0,0 +1,1320 @@
...
+            if state:
+                return ["|", ("state_id", "=", False), ("state_id", "=", state.id)]
+        return False
+
+    @api.model
+    def _match_partner_ref(self, partner_dict, chatter_msg, domain, order):
+        """If a ref is explicitly given, we just want to match that partner"""
+        if partner_dict.get("ref"):
+            return self.env["res.partner"].search(
+                domain + [("ref", "=", partner_dict["ref"])], limit=1, order=order
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Better to build domain like this:
```
from odoo.osv import expression  # Move it on the import secion
domain = expression.AND([domain, [("ref", "=", partner_dict["ref"])]])
self.env["res.partner"].search(domain, limit=1, order=order)
```

File: base_business_document_import/models/business_document_import.py [hunk 4]
```diff
@@ -0,0 +1,1320 @@
...
+            )
+        return False
+
+    @api.model
+    def _match_partner_contact(self, partner_dict, chatter_msg, domain, order):
+        rpo = self.env["res.partner"]
+        if partner_dict.get("email") and "@" in partner_dict["email"]:
+            partner = rpo.search(
+                domain + [("email", "=ilike", partner_dict["email"])],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 5]
```diff
@@ -0,0 +1,1320 @@
...
+                limit=1,
+                order=order,
+            )
+            if partner:
+                return partner
+        if partner_dict.get("contact"):
+            partner = rpo.search(
+                domain + [("name", "=ilike", partner_dict["contact"])],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 6]
```diff
@@ -0,0 +1,1320 @@
...
+            )
+            if partner:
+                return partner
+        if partner_dict.get("phone"):
+            partner = rpo.search(
+                domain
+                + [
+                    "|",
+                    ("mobile", "=", partner_dict["phone"]),
+                    ("phone", "=", partner_dict["phone"]),
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 7]
```diff
@@ -0,0 +1,1320 @@
...
+            )
+            if partner:
+                return partner
+        return False
+
+    @api.model
+    def _match_partner_name(self, partner_dict, chatter_msg, domain, order):
+        if partner_dict.get("name"):
+            return self.env["res.partner"].search(
+                domain + [("name", "=ilike", partner_dict["name"])],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 8]
```diff
@@ -0,0 +1,1320 @@
...
+            netloc = urlp.path
+        if netloc and len(netloc.split(".")) >= 2:
+            return ".".join(netloc.split(".")[-2:])
+
+    @api.model
+    def _match_partner_website(self, partner_dict, chatter_msg, domain, order):
+        website_domain = self._get_partner_website_domain(partner_dict)
+        if website_domain:
+            return self.env["res.partner"].search(
+                domain + [("website", "=ilike", "%" + website_domain + "%")],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 9]
```diff
@@ -0,0 +1,1320 @@
...
+
+    @api.model
+    def _match_partner_email(self, partner_dict, chatter_msg, domain, order):
+        email_domain = self._get_partner_email_domain(partner_dict)
+        # I can't search on email addresses with
+        # email_domain because of the emails such as
+        # @gmail.com, @yahoo.com that may match random partners
+        if email_domain:
+            partner = self.env["res.partner"].search(
+                domain + [("website", "=ilike", "%" + email_domain + "%")],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 10]
```diff
@@ -0,0 +1,1320 @@
...
+                limit=1,
+                order=order,
+            )
+            if not partner:
+                partner = self.env["res.partner"].search(
+                    domain + [("email", "=ilike", "%@" + email_domain)],
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 11]
```diff
@@ -0,0 +1,1320 @@
...
+            domain += country_domain
+
+            state_domain = self._get_country_state_filter(partner_dict, chatter_msg)
+            if state_domain:
+                domain += state_domain
+
+        # Search on VAT
+        if partner_dict.get("vat"):
+            vat = partner_dict["vat"].replace(" ", "").upper()
+            partner = rpo.search(domain + [("vat", "=", vat)], limit=1, order=order)
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 12]
```diff
@@ -0,0 +1,1320 @@
...
+            'street': 'Long Avenue',
+            'street2': 'Building 2A',
+            'city': 'Paris',
+            'zip': '69100',
+            'country_code': 'FR',
+            }
+        The shipping partner can be any partner, not especially related to the
+        customer/supplier (partner argument)
+        """
+        domain = domain or []
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py:429]: Same here for the entire section, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 13]
```diff
@@ -0,0 +1,1320 @@
...
+            return product
+        product = self._match_product_search(product_dict)
+        if product:
+            return product
+        elif seller:
+            # WARNING: Won't work for multi-variant products
+            # because product.supplierinfo is attached to product template
+            sinfo = self.env["product.supplierinfo"].search(
+                self._match_company_domain()
+                + [
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py:665]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 14]
```diff
@@ -0,0 +1,1320 @@
...
+                supplier=seller and seller.name or "",
+            ),
+        )
+
+    @api.model
+    def _match_product_search(self, product_dict):
+        product = self.env["product.product"].browse()
+        cdomain = self._match_company_domain()
+        if product_dict.get("barcode"):
+            domain = cdomain + [
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 15]
```diff
@@ -0,0 +1,1320 @@
...
+                "|",
+                ("barcode", "=", product_dict["barcode"]),
+                ("packaging_ids.barcode", "=", product_dict["barcode"]),
+            ]
+            product = product.search(domain, limit=1)
+        if not product and product_dict.get("code"):
+            # TODO: this domain could be probably included in the former one
+            domain = cdomain + [
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py]: Same here, better to merge domain with `expression.AND(...)`

File: base_business_document_import/models/business_document_import.py [hunk 16]
```diff
@@ -0,0 +1,1320 @@
...
+        if currency_dict.get("iso_or_symbol"):
+            currencies = rco.search(
+                [
+                    "|",
+                    ("name", "=", currency_dict["iso_or_symbol"].upper()),
+                    ("symbol", "=", currency_dict["iso_or_symbol"]),
+                ]
+            )
+            if len(currencies) == 1:
+                return currencies[0]
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py:783]: Useless to specify `[0]`

**Review by acsonefho — CHANGES_REQUESTED** [review]
Some improvements

File: base_business_document_import/models/business_document_import.py (hunk 2 above)
Comment (ghost) [reply base_business_document_import/models/business_document_import.py]: @acsonefho Nice, looks more clear that way

**Comment by ghost** [comment]
@acsonefho Just implemented your feedback changes, can you please review?

File: base_business_document_import/models/business_document_import.py [hunk 17]
```diff
@@ -141,68 +136,86 @@ def _get_country_state_filter(self, partner_dict, chatter_msg):
             country = self.env["res.country"].search(
                 [("code", "=", partner_dict["country_code"])], limit=1
             )
-            state = self.env["res.country.state"].search(
+            if state := self.env["res.country.state"].search(
```
Comment (acsonefho) [inline base_business_document_import/models/business_document_import.py:139]: IMO Not in favor of this syntax who is less readable.. But why not.

**Review by acsonefho — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

**Comment by acsonefho** [comment]
@OCA/edi-maintainers Seems ready to merge? :pray: Thanks in advance :muscle:

**Review by TDu — APPROVED** [review]

**Comment by acsonefho** [comment]
@rousseldenis Do you have super-power here to merge?

**Comment by rousseldenis** [comment]
> @rousseldenis Do you have super-power here to merge?

Nope sorry. @OCA/edi-maintainers

**Comment by sebalix** [comment]
cc @simahawk this one is ready to merge

**Comment by simahawk** [comment]
checking...

**Review by simahawk — APPROVED** [review]
I guess we can live a bit more w/ a deprecated method...

**Pull request merged**